### ISSUE TITLE
Session Timeout Duration

### ISSUE THREAD

**Issue by Lukejkw** [AxaGuilDEv/react-oidc#796]
Hello,

Thanks so much for the library.

Is it possible to extend the session timeout? The session is timing out after a couple minutes when changing tabs. Where can this be configured? I presume the service worker is being paused and causing this?

Thanks

**Comment by guillaume-chervet** [comment]
Serviceworker lifetime is manager by the browser. It xan be killed after 10 second of inactivity. That why I have added a service worker keep alive ping.
Next Ping does not start when tab is unavalable.

I havent tried yet to grow up the ping time. I will make some tests tomorrow to check if it is possible to do it. It woulb be awesome if it works.

On mobile browser are very agressive and kill the serviceworker event if a request is running :(

**Comment by Lukejkw** [comment]
Thanks for the speedy reply. 

Do you have any alternatives you can advise? I'm trying to provide a somewhat connected experience for the user. So when the session times-out, at least redirect the user to the identity provider to reauthenticate and bring them back to where they were.

If I remove the service worker registration and fallback to a less secure mechanism of storing the access token (I presume in local storage) will silent refresh still work? In order to do this, do I just not add the servicer worker and then use hooks to retrieve the access token?

How can I listen for token/session timeout in order to redirect back to the identity provider to reauthenticate when the user returns?I have seen the onSessionLost callback but because this is defined on the OidcProvider, I do not have access to methods on the hooks.

There are a few gaps in the docs prompting these questions. Happy to contribute back to them once I understand the bigger picture.

**Comment by Lukejkw** [comment]
Just to add to this. Testing on mobile (Chrome on iPad), the service worker is very aggressively killed. Even if you are on the page but not using it for a few seconds, the service worker is killed and you need to refresh. There must be a better way of dealing with this?

**Comment by guillaume-chervet** [comment]
It is strange, it should not be as agressive.

I have never lost my service worker during an active tabs thank to the time too live (on android chrome, firefox, edge).
I do not have Ipad or mac, may you did a video of the behavior you have?

I will test to increase it for unfocused tab.
#797

Without service worker I use session storage, localstorage is too unsecure to use it for AXA.

Did you configure silent renew ? On session lost if your have cookies session with your Oidc Provider, silent renew will retrieve your session.

Definitly, if increase TTL does not work, the documentation should be update to explain these behaviors.

**Comment by Lukejkw** [comment]
Yeah, it is unusual how quickly it is killing the service worker. Maybe an iOS thing? The app is proprietary so not sure I can share video here, unfortunately.

I have configured a silent redirect URI but haven't notice a silent refresh happening in the network tab. This is my configuration:

![image](https://user-images.githubusercontent.com/13328460/173024053-948c175d-6d51-4bda-a669-9fa0e3dc2a59.png)

**Comment by omerdrukman** [comment]
Hi all, was there any progress on this one? I'm having the same issue on IOS in Safari and Chrome and also in Safari on Mac (The demo also fails), after authenticating I'm redirected to authentication error following this exception:
![image](https://user-images.githubusercontent.com/15276624/183291906-546df7df-53bd-4a9a-8d27-dc85797d000f.png)

![IMG_1356](https://user-images.githubusercontent.com/15276624/183291495-9009d5bd-1a0f-4f00-9a93-2e68e5ec4119.PNG)
.

On Android it works as expected.

Thanks

**Comment by guillaume-chervet** [comment]
Hi @omerdrukman thank you for the feedback. I am in holiday for 2 weeks. I will ask my collegues who use Mac to test it. It is strange because i have already test with them some month ago. May be an update or specific configuration.

**Comment by guillaume-chervet** [comment]
I have some collegue that can reproduce it on iPhone. I will investigate and i have some ideas. I am in holiday for 2 weeks, i may not be able to do something before.

**Comment by guillaume-chervet** [comment]
Does this version work on your Iphone/ipad? 
https://black-rock-0dc6b0d03-829.westeurope.1.azurestaticapps.net/

**Comment by TheFoolsErrand** [comment]
Hi there @guillaume-chervet I seem to be experiencing a similar issue ("Error: Service Worker storage disapear").

Tested your above version (https://black-rock-0dc6b0d03-829.westeurope.1.azurestaticapps.net/) in:

- Safari Version 15.6 (17613.3.9.1.5) macOS Monterey 12.5 (21G72)
- Safari Version Version 16.0 (18614.1.22.1.2) macOS Ventura 13.0 Beta (22A5311f)
- Safari iPadOS 15.6

Interestingly sometimes it works (on a clean session) but it times out instantly

```
10:36:5 oidc:default:logout_from_another_tab: {"message":"SessionMonitor","sub":"2"}
10:36:5 oidc:default:silentLoginAsync_error: {"error":"login_required"}
10:36:4 oidc:default:silentLoginAsync_begin: {}
10:36:4 oidc:default:token_aquired: xx
```

I also happened to notice if you set `monitor_session` to false and you re-load after the first error it seems to be ok until the token expires.

Possibly related to the vanilla timer.ts 'SharedWorker not available'?

Sounds like Apple is going to make us work hard for this one ;)

**Comment by guillaume-chervet** [comment]
I will need friends to share me a physical Mac/Ipad I think :)

I have disable serviceWorker temporaly for iOS and Mac OS.
https://black-rock-0dc6b0d03-829.westeurope.1.azurestaticapps.net/ => does it work ?

The strange thing is I have V5 in production used by mac OS X at work (and it is working).

**Comment by TheFoolsErrand** [comment]
Yes, I also had v5 working on the same project. So could be something new introduced in v6... perhaps the logout_from_another_tab / session monitor? I noticed it in the first log I posted, but it also popped up this morning with your above:

```
6:15:8 oidc:default:logout_from_another_tab: {"message":"SessionMonitor","sub":"2"}
6:15:8 oidc:default:silentLoginAsync_error: {"error":"login_required"}
6:15:8 oidc:default:token_timer: {"timeLeft":61}
6:15:7 oidc:default:token_timer: {"timeLeft":62}
6:15:6 oidc:default:token_timer: {"timeLeft":63}
6:15:5 oidc:default:silentLoginAsync_begin: {}
6:15:5 oidc:default:token_aquired: {"accessToken":"..."}
6:15:5 oidc:default:loginCallbackAsync_end: {}
6:15:3 oidc:default:loginCallbackAsync_begin: {}
```

![image](https://user-images.githubusercontent.com/25444471/183757531-71eb912a-a696-4bed-ad84-129d814d0a47.png)

**Comment by guillaume-chervet** [comment]
Thank you very much for the feedback @TheFoolsErrand , i set up session mobitor to false here https://black-rock-0dc6b0d03-829.westeurope.1.azurestaticapps.net/
Does it change something for you?

**Comment by TheFoolsErrand** [comment]
It seemed to get further that time. On a clean session. I.e. full cycle with login.
Re-Authenticating the same session gets us back to the 'Error authentication' and "Error: Service Worker storage disapear".

```
7:33:40 oidc:default:refreshTokensAsync_error: {"message":"refresh token silent"}
7:33:40 oidc:default:refreshTokensAsync_silent_error: {"message":"exceptionSilent","exception":"oidc_login_required"}
7:33:40 oidc:default:silentLoginAsync_error: {"error":"login_required"}
7:33:37 oidc:default:silentLoginAsync_begin: {}
7:33:37 oidc:default:refreshTokensAsync_begin: {"refreshToken":"REFRESH_TOKEN_SECURED_BY_OIDC_SERVICE_WORKER_default","status":"REQUIRE_SYNC_TOKENS","tryNumber":0}
7:33:37 oidc:default:token_timer: {"timeLeft":61}
7:33:36 oidc:default:token_timer: {"timeLeft":62}
7:33:35 oidc:default:token_timer: {"timeLeft":63}
7:33:34 oidc:default:token_timer: {"timeLeft":64}
7:33:33 oidc:default:token_renewed: {}
7:33:33 oidc:default:refreshTokensAsync_end: {"success":true}
7:33:32 oidc:default:refreshTokensAsync_begin: {"refreshToken":"REFRESH_TOKEN_SECURED_BY_OIDC_SERVICE_WORKER_default","status":"TOKENS_INVALID","tryNumber":0}
7:33:32 oidc:default:token_timer: {"timeLeft":0}
7:33:31 oidc:default:token_timer: {"timeLeft":1}
...
7:32:28 oidc:default:token_timer: {"timeLeft":64}
7:32:27 oidc:default:token_aquired: {"accessToken":"..."}
7:32:27 oidc:default:loginCallbackAsync_end: {}
7:32:26 oidc:default:loginCallbackAsync_begin: {}
```

**Comment by guillaume-chervet** [comment]
I am back from holiday,
I have published a new version where monitorSession is to false by default.

The probleme come from safari cookies tier blocked by default : 
https://infi.nl/nieuws/spa-necromancy/
https://github.com/IdentityModel/oidc-client-js/issues/1074

I will add a new property to monitor session into the same application that will not use cookie, but it won't be able to disconnect from multiple application on different domain.

**Comment by mrige** [comment]
Hello, 
I am experiencing similar issue with safari on Mac OS. after the login happens, the sessions times out. it wasn't an issue I had before its new. I'm not sure if there is a regression in the latest version.

**Comment by guillaume-chervet** [comment]
Hi @mrige , thank you for the feedback.

Which version did you use before that was working?
Do you use ServiceWorker ?

**Comment by mrige** [comment]
Apologies,
I should clarify, it isn't work on the older version as well which was 6.6.6. I have updated to the latest and I have the same issue.
Yes I use ServiceWorker
thanks for the fast reply

**Comment by guillaume-chervet** [comment]
Thank you very much for your help.
Do you have a sample of your configuration ? (Whithout sensitive data)

**Comment by mrige** [comment]
![Screen Shot 2023-01-04 at 7 14 37 PM](https://user-images.githubusercontent.com/16493533/210673726-dd381fa4-35a9-4031-a984-d5cb72ddffcc.png)

**Comment by guillaume-chervet** [comment]
Thank you very much @mrige . I will try today the lastest version with People who own Mac :)

**Comment by mrige** [comment]
I just realized when I renew the token on chrome I get this error 
{
    "error": "login_required",
    "errorDescription": "End-User authentication is required",
    "state": "R81zbKchi6"
}

**Comment by guillaume-chervet** [comment]
@fdescamps made some tests with today on his Mac using the online demo and it seems to work.
@mrige do you have more details about requests and general behavior you get?

**Comment by guillaume-chervet** [comment]
Are you using identity server?

**Comment by mrige** [comment]
<img width="1910" alt="Screen Shot 2023-01-05 at 3 59 19 PM" src="https://user-images.githubusercontent.com/16493533/210878960-aaaca66e-4ad2-4a46-bf5b-d097df779f1f.png">

**Comment by guillaume-chervet** [comment]
Hi @mrige may you show your network panel?
Do you have an accessToken refreshToken ans idtoken example that you receive as a callback?
Which version do you use exatcly now?

**Comment by mrige** [comment]
Yes I get the tokens
the accesstoken opaque : 
access_token: "0DaW3zFx62SLpwTmmRuu34BmZrrAOGbxKjNiw3GTtGy"
refresh_token : same format 
idToken: jwt

version : "@axa-fr/react-oidc": "^6.11.4-alpha1"

<img width="783" alt="Screen Shot 2023-01-06 at 10 44 08 AM" src="https://user-images.githubusercontent.com/16493533/211047308-50e59565-007e-491f-9e17-157a38d13d51.png">

**Comment by guillaume-chervet** [comment]
Hi @mrige thank you for your answer.

Dont use alpha version, it is a big refactoring and a buggy one.
I can reproduce similary bug on it. Please use stable version. I publish alpha to rest on man devices and to make long test on mobile Phone.

**Comment by guillaume-chervet** [comment]
I @Lukejkw , your issue is a very old one. It think it is working now.
Thank you very much for it and feel free to reopen it if you need it.
